src/coreclr/jit/gentree.cpp

```diff
@@ -8672,8 +8672,7 @@ GenTree* Compiler::gtNewLoadValueNode(var_types type, ClassLayout* layout, GenTr
     {
         unsigned   lclNum = addr->AsLclFld()->GetLclNum();
         LclVarDsc* varDsc = lvaGetDesc(lclNum);
-        if ((varDsc->TypeGet() == type) &&
-            ((type != TYP_STRUCT) || ClassLayout::AreCompatible(layout, varDsc->GetLayout())))
+        if ((varDsc->TypeGet() == type) && ((type != TYP_STRUCT) || layout->CanAssignFrom(varDsc->GetLayout())))
         {
             return gtNewLclvNode(lclNum, type);
```

Comment:
I wonder if this will result in correct GC tracking of the fields in the face of whole-struct enregistration or promotion...

Reply:
Is this one required for your other change? I wonder if we can keep the invariant for `LCL_VAR` and only use the new relation for stores.

Reply:
I don't think it is required, but I haven't rebuilt my other changes on top of this yet.

Reply:
The proposed `CanAssignFrom` won't allow you to lose GC tracking -- if the source is a GC type the dest must be also. It will allow ref->byref, nint->byref.

Reply:
My worries are more along the lines of what happens when the backend sees an enregistered `LCL_VAR VXY` with a `TYP_REF` used in `TYP_BYREF` context, and whether the GC reporting is properly updated to be `TYP_BYREF` if e.g. no physical move ends up being required. I suppose we can go with it and see.